Adding changes to separate unit tests and integration tests

jespino opened the pull request:
#### Summary
This separation is not strictly clear, but we are separating now tests that needs database from tests that doesn't need it, so the CI is going to run all the tests but locally we going to run only the not-database-dependant (quick) tests.

agnivade commented on app/helper_test.go:
```diff
@@ -105,11 +101,35 @@ func setupTestHelper(enterprise bool, tb testing.TB) *TestHelper {
 
 func Setup(tb testing.TB) *TestHelper {
-	return setupTestHelper(false, tb)
+	if testing.Short() {
```
If we follow the logic I mentioned above, then we don't need to split this into a separate UnitSetup function and just use the testing.Short() condition to pass the correct store to `setupTestHelper`. This would also reduce modification of a lot of lines which chance Setup to UnitSetup.

jespino replied:
This is not accidental, Setup is for setting the whole thing up, and UnitSetup is for setting things up but with a mock database. The former is going to be skipped entierly on `-short` flag, the later is going to be execute always. If we always use the mock for all tests depending on the `-short` flag, we need to implement all the mock actions to make it work properly.

This idea here is to keep tests without database needs, or where we don't need integrations tests quick, and keep the other integration tests working, with the less disruption possible.

jespino commented on api4/apitestlib.go:
```diff
@@ -147,18 +145,48 @@ func setupTestHelper(enterprise bool, updateConfig func(*model.Config)) *TestHel
 }
 
-func SetupEnterprise() *TestHelper {
+func SetupEnterprise(tb testing.TB) *TestHelper {
```
we are already using that approach, the thing is, if the user is setting up the environment with `Setup` instead of `UnitSetup` i going to skip the test if you are running with `-short`. Basically what I'm doing here is, if you are going to need to instantiate the real database and the flag `-short` is passed, skip that test. That makes a lot of tests to run faster because doesn't need to clean up the database.

The logic for `UnitSetup` and `Setup` is outside the `testing.Short()`, because is exactly what give as the posibility of distinguish between tests that use database or not, and then allow to skip them automatically.

lieut-data commented on Makefile:
```diff
@@ -261,6 +261,15 @@ else
 endif
 	./scripts/test.sh "$(GO)" "$(GOFLAGS)" "$(ALL_PACKAGES)" "$(TESTS)" "$(TESTFLAGS)" "$(GOBIN)"
 
+unit-test-server: start-docker go-junit-report do-cover-file ## Runs tests.
```
I see we do this for `test-server` already, but I'm curious how `go-junit-report` works here. Since it will get run as a dependency of `unit-test-server` before we ever `$(GO) test`, it seems like it won't have anything on which to generate a report?

lieut-data replied:
Can we tweak the help comment here from `Runs tests` to something that captures the changes here?

jespino replied:
Sure, we can add the help comment.

This change is for run the unit tests locally without any coverage reports and things like that. I'm not 100% sure about the need of this task by itself because you can do it manually.

Anyway, this execution task is executed in like 3 minutes and the one with the coverage is taking 6.

lieut-data commented on Makefile:
```diff
@@ -261,6 +261,15 @@ else
+unit-test-server: start-docker go-junit-report do-cover-file ## Runs tests.
+ifeq ($(BUILD_ENTERPRISE_READY),true)
+	@echo Running all tests
+	$(GO) test $(GOFLAGS) $(TESTFLAGS) $(ALL_PACKAGES)
```
I was expecting a `-short` or something here. It seems like this just runs the full tests anyway?

jespino replied:
nope, the `TESTFLAGS` variable is including the `-short` parameter. Probably we can remove the variable here and use directly the `-short` flag

Pomyk replied:
I think it's not obvious that `test-server` by default uses `-short` and doesn't run all tests.  Also `go-junit-report do-cover-file` are not needed for `unit-test-server`.

crspeller replied:
Should we make it more obvious what is happening? Just have `make test` run all the tests and if you want the quick version you run this target?

jespino replied:
Yes, I think that should be the approach. I'll do it.

lieut-data commented on api4/apitestlib.go:
```diff
@@ -147,18 +145,48 @@ func setupTestHelper(enterprise bool, updateConfig func(*model.Config)) *TestHel
 	return setupTestHelper(false, updateConfig)
 }
 
+func UnitSetupConfig(tb testing.TB, updateConfig func(cfg *model.Config)) *TestHelper {
```
nit: `SetupUnit` than `UnitSetup` might read more naturally. 0/5

lieut-data commented on app/config_test.go:
```diff
@@ -51,20 +51,20 @@ func TestConfigListener(t *testing.T) {
 
 func TestAsymmetricSigningKey(t *testing.T) {
-	th := Setup(t).InitBasic()
+	th := UnitSetup(t)
```
It's not clear to me why this should be a unit test -- we're relying on `ensureAsymmetricSigningKey` in the course of app startup to generate this key correctly, which in turn uses the system store to read/write. I'm guessing the mock store does that now... but how do we know that that the real store still does that correctly?

jespino replied:
The reliability of the store is checked a different layers, one is the store layer, other is the app layer and other is the api layer. We can revert this to use integration tests here.

jespino commented:
> I haven't finished my review, but a few thoughts:
> 
>     * Adding `t` to the `Setup` call is long overdue -- very nice. Though I would have loved to see this broken out into its own commit for ease in reviewing.

I can broke that, but was part of the process of implementing it.

>     * I scrolled down to the first "interesting" change `UnitSetup` and see that we're not copying the old test but just switching to the mock store. More comments inline.

We can copy all test or we can switch tests, it depends on the value provided by the integration tests in that context. Most of the tests that I changed to use unit tests haven't any need of database access, and other use widely commonly used endpoints of the store that are heavily tested by the rest of the application.

>     * Truly "unit tests" can already run without the database by just not invoking `Setup`. So really this is about running the app with a fake backing store instead of the real one. Are we 100% confident this won't introduce a potential for regression? The cost to benefit ratio seems very low here, especially if we're trading potential for incorrectness just to speed up developer tests. If we make some headway on just speeding up the database as being discussed, is the meat of this change still worthwhile?

In my opinion, the changes proposed here are pretty secure, when we change to the unit tests is going to use a mock that is only design to provide the methods needed to start up the store, that means the other tests are using that methods or are not using the store at all.

There are specific exceptions where i added mocks implementations to test concrete cases (notification emails testing)

> 
> To tweak the proposal here, would we consider a change to the `UnitSetup*` methods to /only/ use the mock store for `-short`, and otherwise fall through to the normal store? Then we can guarantee the existing behaviour in CI, but expose a shortcut method for developers who want a quick smoketest.

I don't see the value of forcing a real implementation of the store on tests that are not using the store itself, or are using methods that we consider enough tested anyway. If we start to follow the path of mocking the store everywhere, I can understand that approach, but is not the current idea.

jespino commented on api4/apitestlib.go:
```diff
@@ -51,19 +52,13 @@ type TestHelper struct {
+var mainHelper *testlib.MainHelper
 
+func setupTestHelper(dbStore store.Store, enterprise bool, updateConfig func(*model.Config)) *TestHelper {
+	if testing.Short() {
```
In theory, the current approach doesn't allow any "short tests" to be executed using the database, so there is no need of DropAllTables. Anyway, I think now we can remove that and relay on the mocking approach, so we can mock that 2 methods, and depending on your store implementation is going to have any effect or not.

jespino commented on api4/cors_test.go:
```diff
@@ -119,7 +119,7 @@ func TestCORSRequestHandling(t *testing.T) {
 	} {
 		t.Run(name, func(t *testing.T) {
-			th := SetupConfig(func(cfg *model.Config) {
+			th := UnitSetupConfig(t, func(cfg *model.Config) {
```
The idea of Unit test vs Integration tests now is about the usage of the real store or a mock store, if the logic that you need to test is mainly related to the store, you should be doing integration tests, if the logic that you want to test is accidentally related to the store, probably you want to use the unit test and mock the store parts that you need to simulate.

jespino commented on app/helper_test.go:
```diff
@@ -105,11 +101,35 @@ func setupTestHelper(enterprise bool, tb testing.TB) *TestHelper {
+	if testing.Short() {
+		tb.SkipNow()
+	}
+	dbStore := mainHelper.GetStore()
```
There were different from the beginning, probably wasn't intentional, we can replicate this approach here too.

jespino commented on app/busy_test.go:
```diff
@@ -23,10 +23,10 @@ func TestBusySet(t *testing.T) {
 	require.False(t, busy.IsBusy())
 
-	busy.Set(time.Second * 3)
+	busy.Set(time.Millisecond * 100)
```
Yes, the idea is not going to generate problems because we are checking with `require.Eventually` there is not point in waiting 3 seconds

jespino commented on app/authorization_test.go:
```diff
@@ -4,14 +4,15 @@
 
 func TestCheckIfRolesGrantPermission(t *testing.T) {
-	th := Setup(t).InitBasic()
+	th := Setup(t)
```
We didn't need any "initial data" to test the permissions and roles, we are adding this `InitBasic` method everywhere, but only is needed if you going to use teams, channels and users created during that `InitBasic` method.

jespino commented:
@crspeller this is waiting for your review, feel free to remove yourself from the review if you want, there are already 2 developers review and only affects tests, so it doesn't need QA pass.
